### PR TITLE
react native 0.59 and move to react-native-mapbox-gl/maps

### PR DESCRIPTION
**Pull request by geohacker**
This PR addresses #1 #33.

It still needs work:
* [ ] Setup react-native-mapbox-gl on iOS
* [ ] Inspect why android is crashing

cc @sethvincent @batpad

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -84,24 +84,25 @@
     "xmldom": "^0.1.27"
   },
   "devDependencies": {
-    "babel-core": "^7.0.0-bridge.0",
+    "@babel/runtime": "^7.4.3",
+    "babel-core": "^7.4.3",
```
Comment (sethvincent) [inline package.json]: They changed the package naming scheme (I think as part of 7.0.0 changes) so this should be @babel/core.

```suggestion
    "@babel/core": "^7.4.3",
```

**Comment by sethvincent** [comment]
When the crash happens on android I'm seeing this on logcat:

```
07-30 11:01:02.627 20892-20892/com.developmentseed.observe E/AndroidRuntime: FATAL EXCEPTION: main
    Process: com.developmentseed.observe, PID: 20892
    java.lang.ExceptionInInitializerError
        at com.mapbox.mapboxsdk.module.http.HttpRequestImpl.setOkHttpClient(HttpRequestImpl.java:103)
        at com.mapbox.mapboxsdk.module.http.HttpRequestUtil.setOkHttpClient(HttpRequestUtil.java:50)
        at com.developmentseed.observe.MainActivity.onCreate(MainActivity.java:95)
        at android.app.Activity.performCreate(Activity.java:7009)
        at android.app.Activity.performCreate(Activity.java:7000)
        at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1214)
        at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:2731)
        at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:2856)
        at android.app.ActivityThread.-wrap11(Unknown Source:0)
        at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1589)
        at android.os.Handler.dispatchMessage(Handler.java:106)
        at android.os.Looper.loop(Looper.java:164)
        at android.app.ActivityThread.main(ActivityThread.java:6494)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:438)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:807)
     Caused by: com.mapbox.mapboxsdk.exceptions.MapboxConfigurationException: 
    Using MapView requires calling Mapbox.getInstance(Context context, String accessToken) before inflating or creating the view. The access token parameter is required when using a Mapbox service.
    Please see https://www.mapbox.com/help/create-api-access-token/ to learn how to create one.
    More information in this guide https://www.mapbox.com/help/first-steps-android-sdk/#access-tokens.
        at com.mapbox.mapboxsdk.Mapbox.validateMapbox(Mapbox.java:178)
        at com.mapbox.mapboxsdk.Mapbox.getApplicationContext(Mapbox.java:110)
        at com.mapbox.mapboxsdk.http.HttpIdentifier.getIdentifier(HttpIdentifier.java:21)
        at com.mapbox.mapboxsdk.module.http.HttpRequestImpl.<clinit>(HttpRequestImpl.java:39)
```

**Comment by sethvincent** [comment]
Looks like there's an `addCustomHeader` method in react-native-mapbox-gl/maps that might be incompatible with the way we set the http client in our MainActivity.java `onCreate` method: https://github.com/react-native-mapbox-gl/maps/blame/59238cd3c9a21cbc875989d7efcec7ecc7a9db74/android/rctmgl/src/main/java/com/mapbox/rctmgl/modules/RCTMGLModule.java#L308

That's my best guess so far about why the crash is happening.

**Comment by sethvincent** [comment]
We'll need to update MapboxGL.StyleSheet usage to match the breaking changes: https://github.com/react-native-mapbox-gl/maps/blob/master/docs/StyleSheet.md

**Comment by sethvincent** [comment]
I've converted our use of `MapboxGL.StyleSheet` to style expressions based on the guide I linked in the last comment. I _think_ it all works, but it still needs to be tested properly after resolving the `addCustomHeader` incompatibility.

**Comment by sethvincent** [comment]
We'll need to start using the `Camera` component. It takes props like `zoomLevel`, `centerCoordinates`, `followUserMode` (instead of `userTrackingMode`), etc.

https://github.com/react-native-mapbox-gl/maps/blob/master/docs/Camera.md

**Comment by sethvincent** [comment]
I verified that there is some kind of incompatibility between the `addCustomHeader` method in react-native-mapbox-gl/maps and the way we set the http client in our MainActivity.java `onCreate` method by commenting out the http client code: https://github.com/developmentseed/observe/blob/master/android/app/src/main/java/com/developmentseed/observe/MainActivity.java#L65-L100

I'm not sure what a fix would look like yet though.

**Comment by geohacker** [comment]
I looked at this a bit more and think it comes down to the way we're override the HTTP client with one we create to add the cookies. https://github.com/developmentseed/observe/blob/master/android/app/src/main/java/com/developmentseed/observe/MainActivity.java#L95-L101

Don't have any ideas on a fix yet.

**Comment by geohacker** [comment]
`addCustomHeader` likely doesn't have anything to do with what we're running into because we're using 7.0.0.rc-3 and it doesn't have that particular change. https://github.com/react-native-mapbox-gl/maps/compare/7.0.0-rc3...master

`addCustomHeader` is only in master.

**Comment by geohacker** [comment]
Couple of more things I tried:
* change mapbox-sdk and okhttp to match https://github.com/react-native-mapbox-gl/maps/blob/59238cd3c9a21cbc875989d7efcec7ecc7a9db74/android/rctmgl/build.gradle#L26
* look at cookie definition 

Starting to think that by setting httpclient we may be losing the context so we get the accessToken error?

**Comment by geohacker** [comment]
Perhaps related https://github.com/mapbox/mapbox-gl-native/issues/13253

**Comment by geohacker** [comment]
@sethvincent I inspected this a bit more along with the comments from rctmgl and mapbox gl native developers managed to create the instance with the access token before overriding the http client. This doesn't crash anymore. 

Also, previously when we were fetching the context that was coming out to be null because JS modules weren't loaded yet. So this waits for the context to be ready before using it as [explained here](https://stackoverflow.com/a/49176645).

**Comment by sethvincent** [comment]
It took a while to get things running.

I added the `<MapboxGL.UserLocation />` component to see if that helped somehow with user location tracking. I don't think it did.

It seems like the layer, locate user, and add feature buttons, basically all the ui elements other than the map, are not responding to clicks.

If I try to log from those onPress handlers I never see anything logged.

This strikes me as possibly a react-native problem rather than a mapbox-gl problem.

In android studio the mapview is constantly spamming logs like this:

```
08-07 17:31:58.190 7257-7257/com.developmentseed.observe D/MOVE_EVENT: onCameraMoveStarted SENDING WILL_CHANGE EVENT reason: 3 isUserInteraction: false isAnimated: true
    onCameraIdle SENDING DID_CHANGE EVENT isUserInteraction: false isAnimated: true
08-07 17:31:58.196 7257-7257/com.developmentseed.observe D/MOVE_EVENT: onCameraMoveStarted SENDING WILL_CHANGE EVENT reason: 3 isUserInteraction: false isAnimated: true
    onCameraIdle SENDING DID_CHANGE EVENT isUserInteraction: false isAnimated: true
08-07 17:31:58.315 7257-7257/com.developmentseed.observe D/MOVE_EVENT: onCameraMoveStarted SENDING WILL_CHANGE EVENT reason: 3 isUserInteraction: false isAnimated: true
    onCameraIdle SENDING DID_CHANGE EVENT isUserInteraction: false isAnimated: true
08-07 17:31:58.318 7257-7257/com.developmentseed.observe D/MOVE_EVENT: onCameraMoveStarted SENDING WILL_CHANGE EVENT reason: 3 isUserInteraction: false isAnimated: true
    onCameraIdle SENDING DID_CHANGE EVENT isUserInteraction: false isAnimated: true
08-07 17:31:58.449 7257-7257/com.developmentseed.observe D/MOVE_EVENT: onCameraMoveStarted SENDING WILL_CHANGE EVENT reason: 3 isUserInteraction: false isAnimated: true
    onCameraIdle SENDING DID_CHANGE EVENT isUserInteraction: false isAnimated: true
08-07 17:31:58.554 7257-7257/com.developmentseed.observe D/MOVE_EVENT: onCameraMoveStarted SENDING WILL_CHANGE EVENT reason: 3 isUserInteraction: false isAnimated: true
    onCameraIdle SENDING DID_CHANGE EVENT isUserInteraction: false isAnimated: true
08-07 17:31:58.663 7257-7257/com.developmentseed.observe D/MOVE_EVENT: onCameraMoveStarted SENDING WILL_CHANGE EVENT reason: 3 isUserInteraction: false isAnimated: true
    onCameraIdle SENDING DID_CHANGE EVENT isUserInteraction: false isAnimated: true
```

I think that's somewhat expected as it is trying to follow the user location. If I disable location tracking those logs go away.

Once I ended up seeing this when trying to log the camera ref in the ref callback where we're setting it:

```
'camera ref', null
08-07 17:08:10.972  6361  6912 E ReactNativeJS: Error: Out of memory
08-07 17:08:10.972  6361  6912 E ReactNativeJS: 
08-07 17:08:10.972  6361  6912 E ReactNativeJS: This error is located at:
08-07 17:08:10.972  6361  6912 E ReactNativeJS:     in Camera (at Explore.js:408)
08-07 17:08:10.972  6361  6912 E ReactNativeJS:     in RCTMGLAndroidTextureMapView (at MapView.js:724)
08-07 17:08:10.972  6361  6912 E ReactNativeJS:     in RCTView (at View.js:45)
08-07 17:08:10.972  6361  6912 E ReactNativeJS:     in View (at MapView.js:737)
08-07 17:08:10.972  6361  6912 E ReactNativeJS:     in MapView (at Explore.js:391)
08-07 17:08:10.972  6361  6912 E ReactNativeJS:     in RCTView (at View.js:45)
08-07 17:08:10.972  6361  6912 E ReactNativeJS:     in View (created by Context.Consumer)

...

```

Haven't been able to debug that yet, though. I'm using an emulator currently so there's a chance it can be resolved by changing emulator settings.

Running the app in the emulator can slow down my macbook pro to the point where i can't move the mouse. I haven't been able to try on an android device but I'll do that tomorrow.

Some other seemingly unrelated issues:

I noticed that sometimes the app seems to think it is offline even though it is successfully making requests:
Screen Shot 2019-08-07 at 4 41 22 PM

There are some errors with the map styles:

```
08-07 17:27:53.509 7257-7257/com.developmentseed.observe E/Mbgl: {entseed.observe}[JNI]: Error setting property: line-color [2]: Could not parse color from value 'primary'
08-07 17:27:53.516 7257-7257/com.developmentseed.observe E/Mbgl: {entseed.observe}[JNI]: Error setting property: line-color [2]: Could not parse color from value 'foot'
```

```
08-07 17:53:40.256 8301-8301/com.developmentseed.observe E/Mbgl: {entseed.observe}[JNI]: Error setting property: line-width Expected an even number of arguments.
08-07 17:53:40.259 8301-8301/com.developmentseed.observe E/Mbgl: {entseed.observe}[JNI]: Error setting property: line-color Expected an even number of arguments.
    {entseed.observe}[JNI]: Error setting property: line-width Expected an even number of arguments.
08-07 17:53:40.261 8301-8301/com.developmentseed.observe E/Mbgl: {entseed.observe}[JNI]: Error setting property: line-color Expected an even number of arguments.
08-07 17:53:40.263 8301-8301/com.developmentseed.observe E/Mbgl: {entseed.observe}[JNI]: Error setting property: line-width Expected an even number of arguments.
08-07 17:53:40.394 8301-8301/com.developmentseed.observe E/Mbgl: {entseed.observe}[JNI]: Error setting property: line-width Expected an even number of arguments.
```

Issues relevant to the map style errors:

https://github.com/mapbox/mapbox-gl-js/issues/6434
https://github.com/mapbox/mapbox-gl-js/issues/6437
https://github.com/mapbox/mapbox-gl-js/issues/8446

**Comment by geohacker** [comment]
We're abandoning this update process temporarily. This PR and the ios one #86 has more background on why this is not a good time for us to do the updates -- v0.7.0 mapbox-gl isn't stable enough just yet and also lack some documentation which we are unable to help out with at the moment. 

We'll revisit in a few weeks.

**Pull request closed without merging**